acrg/name/name.py

```diff
@@ -1562,30 +1563,34 @@ def local_solar_time(dataset):
         hours = dataset.time.to_pandas().index.hour
         return hours
     
-    def local_ratio(dataset):
+    @numba.jit()
+    def assign_localness(dataset,wh_rlat,wh_rlon):
+        local_sum=np.zeros(dataset.shape[2])
+        for ti in range(dataset.shape[2]):
+            fp_data_H_site_local = dataset[wh_rlat[0]-2:wh_rlat[0]+3,
+                                                     wh_rlon[0]-2:wh_rlon[0]+3,ti]
+            local_sum[ti] = np.sum(fp_data_H_site_local)/np.sum(dataset[:,:,ti])
+        return local_sum
+
+    def define_localness(fp_data_H,site):
         """
-        Calculates the local ratio in the surrounding grid cells
+        Define the localness of each time point for each site.
+        Sum up the 25 grid boxes surrounding the site.
         """
-        release_lons = dataset.release_lon[0].values
-        release_lats = dataset.release_lat[0].values
-        dlon = dataset.lon[1].values - dataset.lon[0].values
-        dlat = dataset.lat[1].values-dataset.lat[0].values
-        local_sum=np.zeros((len(dataset.mf)))
-
-        for ti in range(len(dataset.mf)):
-            release_lon=dataset.release_lon[ti].values
-            release_lat=dataset.release_lat[ti].values
-            wh_rlon = np.where(abs(dataset.lon.values-release_lon) < dlon/2.)
-            wh_rlat = np.where(abs(dataset.lat.values-release_lat) < dlat/2.)
-            if np.any(wh_rlon[0]) and np.any(wh_rlat[0]):
-                local_sum[ti] = np.sum(dataset.fp[wh_rlat[0][0]-2:wh_rlat[0][0]+3,wh_rlon[0][0]-2:wh_rlon[0][0]+3,ti].values)/\
-                                np.sum(dataset.fp[:,:,ti].values)
-            else:
-                local_sum[ti] = 0.0 
-        
-        return local_sum
-    
-    
+        release_lon=fp_data_H[site].release_lon[0].values
+        release_lat=fp_data_H[site].release_lat[0].values
+        dlon=fp_data_H[site].lon[1].values-fp_data_H[site].lon[0].values
```

Comment:
Small suggestion: you can use the [`.diff(dim=...)` method](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.diff.html) to calculate the difference between values. You could also use the `.mean()` on this to get one value. I just mention this as I know sometimes the [1] - [0] element can sometimes contain mistakes.

This is just a suggestion though - if speed is the limiting factor the solution already included here will likely be quicker.